commons/src/main/java/org/infinispan/commons/api/functional/MetaParam.java

```diff
@@ -0,0 +1,230 @@
+package org.infinispan.commons.api.functional;
+
+import java.util.NoSuchElementException;
+import java.util.Optional;
+
+/**
+ * An easily extensible metadata parameter that's stored along with the value
+ * in the the functional map.
+ *
+ * Some metadata parameters can be provided by the user in which case they
+ * need to implement {@link MetaParam.Writable}. Examples of writable metadata
+ * parameters are version information, lifespan of the cached value...etc.
+ *
+ * Those metadata parameters not extending {@link MetaParam.Writable} are
+ * created by internal logic and hence can only be queried, for example:
+ * time when value was added into the functional map, or last time value
+ * was accessed or modified.
+ *
+ * What makes {@link MetaParam} different from {@link Param} is that {@link MetaParam}
+ * values are designed to be stored along with key/value pairs in the functional map,
+ * to provide extra information. On the other hand, {@link Param} instances
+ * merely act as ways to tweak how operations are executed, and their contents
+ * are never stored permanently.
+ *
+ * DESIGN RATIONALES:
+ * <ul>
+ *    <il>This interface replaces Infinispan's Metadata interface providing
+ *    a more flexible way to add new metadata parameters to be stored with
+ *    the cached entries.
+ *    </il>
+ *    <il>Another benefit of the design is that it makes a clear separation
+ *    between the metadata that can be provided by the user on per-entry basis,
+ *    e.g. lifespan, maxIdle, version...etc, versus metadata that's produced
+ *    by the internal logic that cannot be modified directly by the user, e.g.
+ *    cache entry created time, last time cache entry was modified or accessed
+ *    ...etc.
+ *    </il>
+ * </ul>
+ *
+ * @param <T> type of MetaParam instance, implementations should assign it to
+ * the implementation's type.
+ * @since 8.0
+ */
+public interface MetaParam<T> {
+
+   /**
+    * Provides metadata parameter lookup capabilities using {@link Class} as
+    * lookup key.
+    *
+    * DESIGN RATIONALES:
+    * <ul>
+    *    <li>Why have both getMetaParam() and findMetaParam()? Convenience,
+    *    for exactly the same reasons why {@link EntryView.ReadEntryView}
+    *    exposes {@link EntryView.ReadEntryView#get()} and
+    *    {@link EntryView.ReadEntryView#find()}
+    *    </li>
+    * </ul>
+    */
+   interface Lookup {
+      /**
+       * Returns a non-null metadata parameter implementation of the can be
+       * assigned to the type {@link Class} being looked up, or throws
+       * {@link NoSuchElementException} if no metadata parameter exists.
+       *
+       * @throws NoSuchElementException if no metadata parameter exists.
+       *
+       * @param <T> metadata parameter type
+       */
+      <T> T getMetaParam(Class<T> type) throws NoSuchElementException;
```

Comment:
That's a good point, the REST server might need to double-check for compatibility reasons, but there are ways around it. You could have a layer in place for when compatibility is enabled, and another for when it's not. That's not the case today, but could be done that way. 

An end user though might add some metadata and never use compatibility, I think they'd find this useful. I'm not hugely fuzzed about this...